### ISSUE TITLE
Callback not invoked for the validation set with DDP

### ISSUE THREAD

**Issue by athn-nik** [Lightning-AI/lightning#15028]
### Bug description

I have a callback that is supposed to called during training and validation set. However, the validation part of it is never invoked. The callback is about rendering videos and logging them in wandb. The callback is supposed to happened every n epochs after the train/val step ends. The callback is never called since the videos which are supposed to create are never saved on the disk, so not wandb but pl bug. This happens  when using Gpus > 1 with ddp.

There are no error messages. The videos are not created and none of the print statements indicate that this part of the callback is ever accessed.

### How to reproduce the bug

```python
The code of the callback:
 


class RenderCallback(Callback):
    def __init__(self, bm_path: str = None,
                 path: str = "visuals",
                 logger_type: str = "wandb",
                 save_last: bool = True,
                 vid_format: str = "mp4",
                 every_n_epochs: int = 20,
                 num_workers: int = 0,
                 nvids_to_save: int = 5,
                 fps: float = 30.0,
                 modelname = 'space') -> None:

        if logger_type == "wandb":
            self.log_to_logger = log_to_wandb
        elif logger_type == "tensorboard":
            self.log_to_logger = log_to_tensorboard
        elif logger_type == "none":
            self.log_to_logger = log_to_none
        else:
            raise NotImplementedError("This logger is unknown, please use tensorboard or wandb.")

        self.logger_type = logger_type
        self.path = Path(path)
        self.path.mkdir(exist_ok=True)

        self.fps = fps
        self.nvids = nvids_to_save
        self.every_n_epochs = every_n_epochs
        self.num_workers = num_workers
        self.vid_format = vid_format
        self.save_last = save_last
        self.model = modelname

        from hydra.utils import get_original_cwd
        # self.labels_dict = read_json(f'{get_original_cwd()}/deps/inference/labels.json')
        self.labels_train = read_json(f'{get_original_cwd()}/deps/inference/labels_train_spatial.json')
        self.labels_val = read_json(f'{get_original_cwd()}/deps/inference/labels_val_spatial.json')

        if bm_path is not None:
            self.body_model_path = Path(bm_path) / 'smpl_models'

    def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule,
                           **kwargs) -> None:
        if trainer.is_global_zero:
            return self.call_renderer("train", trainer, pl_module)

    def on_validation_epoch_end(self, trainer: Trainer,
                                pl_module: LightningModule) -> None:
        if trainer.is_global_zero:
            return self.call_renderer("val", trainer, pl_module)

    def on_test_epoch_end(self, trainer: Trainer,
                          pl_module: LightningModule) -> None:
        return self.call_renderer("test", trainer, pl_module)

    def call_renderer(self, split: str, trainer: Trainer,
                      pl_module: LightningModule) -> None:
        if trainer.sanity_checking:
            return

        if self.nvids is None or self.nvids == 0:
            return

        # mid-epoch starting for finetuning
        # if pl_module.store_examples[split] is None:
        #     return
        
        logger.debug(f"Render {split} samples and log to {self.logger_type}")

        # Don't log epoch 0
        if trainer.current_epoch == 0 or trainer.current_epoch % self.every_n_epochs != 0:
            # Log last one (return = don't log, if it is not the last one)
            if trainer.current_epoch != (trainer.max_epochs - 1):
                return
            # Don't log last one if we don't want it
            elif not self.save_last:
                return
        # Prepare the folder
        folder = "epoch_" + str(trainer.current_epoch).zfill(3)
        folder = self.path / folder
        folder.mkdir(exist_ok=True)

        # Extract the stored data
        store_examples = pl_module.store_examples[split]
        ref_joints_or_verts = store_examples['ref']
        ref_motion_features = store_examples['ref_features']
        keyids_to_render = store_examples['keyids']
        ref_motion_features = ref_motion_features.features
        ref_motion_features = ref_motion_features[:self.nvids]

        # Render + log
        # nvids = min(self.nvids, len(ref_joints_or_verts))

        pl_module.eval()
        if split == 'train': 
            render_list = [self.labels_train[key] for key in keyids_to_render]
            # render_list = [lens_acts for key, lens_acts in self.labels_train.items() if key in keyids_to_render]

            texts = [[t0,t1] for t0, t1, _ in render_list]
            lens = [l for _, _, l in render_list]

            jts_T = pl_module.forward_seq(texts, lens, return_type='joints')
            jts_T = [mot.detach().cpu().numpy() for mot in jts_T]

            jts_M = pl_module.forward_motion(ref_motion_features, lens, inference=True, return_type='joints')
            jts_M = [mot.detach().cpu().numpy() for mot in jts_M]

            texts = [f'{t0},{t1} | {keyids_to_render[i]}' for i, (t0, t1) in enumerate(texts)] 

        
        elif split == 'val':
            render_list = [self.labels_val[key] for key in keyids_to_render]

            # render_list = [lens_acts for key, lens_acts in self.labels_val.items() if key in keyids_to_render]
            texts = [[t0,t1] for t0, t1, _ in render_list]

            lens = [l for _, _, l in render_list]

            jts_T = pl_module.forward_seq(texts, lens, return_type='joints')
            jts_T = [mot.detach().cpu().numpy() for mot in jts_T]

            jts_M = pl_module.forward_motion(ref_motion_features, lens, inference=True, return_type='joints')
            jts_M = [mot.detach().cpu().numpy() for mot in jts_M]

            texts = [f'{t0},{t1} | {keyids_to_render[i]}' for i, (t0, t1) in enumerate(texts)] 

        # self.labels_val 
        # render_list_train = list(self.labels_train.values())[:self.nvids]
        # render_list_val =  list(self.labels_val.values())[:self.nvids]
        
        # for set_name, keyids in zip(['train', 'val'], [render_list_train, render_list_val]):
        #     for keyid in keyids:

        import multiprocessing
        list_of_logs = []
        num_workers = min(self.num_workers, 3 * self.nvids)
        with multiprocessing.Pool(num_workers) as pool:
            iterable = ((joints[index], name, index, split,
                         folder, self.fps, description, trainer.current_epoch)
                        for joints, name in zip([jts_T, jts_M, ref_joints_or_verts],
                                                ['text', 'motion', 'ref'])
                        for index, description in zip(range(self.nvids), texts))
            for output, fig_number, name, desc in pool.imap_unordered(render_and_save, iterable):
                split_set = desc.split('_')[-1]
                log_name = f"visuals/{name}/{split_set}_{fig_number}"
                list_of_logs.append((output, log_name, desc))

                train_logger = pl_module.logger.experiment

                # self.log_to_logger(path=output, log_name=log_name, caption=desc,
                #                    fps=self.fps, global_step=trainer.current_epoch,
                #                    train_logger=train_logger, vid_format=self.vid_format)
        import operator
        list_of_logs.sort(key=operator.itemgetter(2))
        for vid_path, panel_name, text_desc in list_of_logs: 
            log_name_start, branch_or_gt, _ = panel_name.split('/')
            vid_id = vid_path.split('/')[-1].split('_')[-1].split('.')[0]
            log_name = f'{log_name_start}_{split}/sample-{vid_id}/{branch_or_gt}'
            self.log_to_logger(path=vid_path, log_name=log_name, caption=text_desc,
                                fps=self.fps, global_step=trainer.current_epoch,
                                train_logger=train_logger, vid_format=self.vid_format)
```

### Error messages and logs

There are no error messages, the code is just never invoked.

### Environment

```

#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):

```

### More info

* Lightning:
	- pytorch-lightning: 1.7.6
	- torch:             1.12.1+cu113
	- torchaudio:        0.12.1+cu113
	- torchmetrics:      0.9.3
	- torchvision:       0.13.1+cu113
* Packages:
	- absl-py:           1.0.0
	- aiohttp:           3.8.1
	- aiosignal:         1.2.0
	- antlr4-python3-runtime: 4.8
	- astroid:           2.12.10
	- asttokens:         2.0.5
	- async-timeout:     4.0.2
	- attrs:             21.4.0
	- autopep8:          1.7.0
	- backcall:          0.2.0
	- cachetools:        5.0.0
	- certifi:           2021.10.8
	- charset-normalizer: 2.0.12
	- chumpy:            0.70
	- click:             8.0.4
	- colorlog:          6.6.0
	- contourpy:         1.0.5
	- cycler:            0.11.0
	- decorator:         4.4.2
	- dill:              0.3.5.1
	- docker-pycreds:    0.4.0
	- einops:            0.4.1
	- et-xmlfile:        1.1.0
	- executing:         0.8.3
	- filelock:          3.6.0
	- fonttools:         4.37.3
	- freetype-py:       2.2.0
	- frozenlist:        1.3.0
	- fsspec:            2022.2.0
	- future:            0.18.2
	- gitdb:             4.0.9
	- gitpython:         3.1.27
	- google-auth:       2.6.2
	- google-auth-oauthlib: 0.4.6
	- grpcio:            1.44.0
	- huggingface-hub:   0.4.0
	- hydra-colorlog:    1.1.0
	- hydra-core:        1.1.1
	- idna:              3.3
	- imageio:           2.16.1
	- imageio-ffmpeg:    0.4.5
	- importlib-metadata: 4.11.3
	- ipdb:              0.13.9
	- ipython:           8.1.1
	- isort:             5.10.1
	- jedi:              0.18.1
	- joblib:            1.1.0
	- kiwisolver:        1.4.4
	- lazy-object-proxy: 1.7.1
	- loguru:            0.6.0
	- markdown:          3.3.6
	- matplotlib:        3.6.0
	- matplotlib-inline: 0.1.3
	- mccabe:            0.7.0
	- more-itertools:    8.14.0
	- moviepy:           1.0.3
	- multidict:         6.0.2
	- networkx:          2.7.1
	- numpy:             1.23.3
	- oauthlib:          3.2.0
	- omegaconf:         2.1.1
	- openai:            0.23.0
	- opencv-python:     4.5.5.64
	- openpyxl:          3.0.10
	- packaging:         21.3
	- pandas:            1.4.1
	- pandas-stubs:      1.4.4.220919
	- parso:             0.8.3
	- pathtools:         0.1.2
	- pexpect:           4.8.0
	- pickleshare:       0.7.5
	- pillow:            9.0.1
	- pip:               22.2.2
	- platformdirs:      2.5.2
	- proglog:           0.1.9
	- promise:           2.3
	- prompt-toolkit:    3.0.28
	- protobuf:          3.19.4
	- psutil:            5.9.0
	- ptyprocess:        0.7.0
	- pudb:              2022.1.1
	- pure-eval:         0.2.2
	- pyasn1:            0.4.8
	- pyasn1-modules:    0.2.8
	- pycodestyle:       2.9.1
	- pydeprecate:       0.3.1
	- pyglet:            1.5.23
	- pygments:          2.11.2
	- pylint:            2.15.3
	- pyopengl:          3.1.0
	- pyparsing:         3.0.7
	- pyrender:          0.1.45
	- python-dateutil:   2.8.2
	- pytorch-lightning: 1.7.6
	- pytz:              2021.3
	- pyyaml:            6.0
	- regex:             2022.3.15
	- requests:          2.27.1
	- requests-oauthlib: 1.3.1
	- rsa:               4.8
	- sacremoses:        0.0.49
	- scipy:             1.8.0
	- sentry-sdk:        1.5.8
	- setproctitle:      1.2.2
	- setuptools:        59.5.0
	- shortuuid:         1.0.8
	- six:               1.16.0
	- smmap:             5.0.0
	- smplx:             0.1.28
	- stack-data:        0.2.0
	- tensorboard:       2.10.0
	- tensorboard-data-server: 0.6.1
	- tensorboard-plugin-wit: 1.8.1
	- termcolor:         1.1.0
	- tokenizers:        0.11.6
	- toml:              0.10.2
	- tomli:             2.0.1
	- tomlkit:           0.11.4
	- torch:             1.12.1+cu113
	- torchaudio:        0.12.1+cu113
	- torchmetrics:      0.9.3
	- torchvision:       0.13.1+cu113
	- tqdm:              4.63.0
	- traitlets:         5.1.1
	- transformers:      4.17.0
	- trimesh:           3.10.5
	- types-pytz:        2022.2.1.0
	- typing-extensions: 4.1.1
	- urllib3:           1.26.9
	- urwid:             2.1.2
	- urwid-readline:    0.13
	- uuid:              1.30
	- wandb:             0.12.11
	- wcwidth:           0.2.5
	- werkzeug:          2.0.3
	- wheel:             0.37.1
	- wrapt:             1.14.1
	- yarl:              1.7.2
	- yaspin:            2.1.0
	- zipp:              3.7.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.14

**Comment by awaelchli** [comment]
Are you passing the dataloader correctly? Like

`trainer.fit(model, train_dataloader, val_dataloader)`

**Comment by athn-nik** [comment]
Yes, I do this and the dataloader is implemented via LightningDataModule. The callback is not called as it does not enter the `validation_on_epoch_end` method of the callback. It does so when I use a single GPU though, and also the same method is normally accesed for the base model attached above.

**Comment by awaelchli** [comment]
@athn-nik I cannot reproduce this. Here is a runnable example based on your configuration (but I removed all code that was either irrelevant or was incomplete for me to use). 

```python
import os

import torch
from torch.optim import Adam
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, Callback


class RenderCallback(Callback):


    def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule,
                           **kwargs) -> None:
        if trainer.is_global_zero:
            print("on train epoch end in callback")

    def on_validation_epoch_end(self, trainer: Trainer,
                                pl_module: LightningModule) -> None:
        if trainer.is_global_zero:
            # return self.call_renderer("val", trainer, pl_module)
            print("on val epoch end in callback")

    def on_test_epoch_end(self, trainer: Trainer,
                          pl_module: LightningModule) -> None:
        # return self.call_renderer("test", trainer, pl_module)
        print("on test epoch end in callback")

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return Adam(self.parameters())

    def training_epoch_end(self, outputs):
        print("train epoch end on epoch", self.current_epoch)

    def validation_epoch_end(self, outputs):
        print("val epoch end on epoch", self.current_epoch)

    def test_epoch_end(self, outputs):
        print("test epoch end")


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        # auto_select_gpus=True
        strategy="ddp",
        benchmark=False,
        max_epochs=1001,
        accelerator="cpu",
        devices=2,
        log_every_n_steps=1,
        deterministic=False,
        detect_anomaly=False,
        enable_progress_bar=False,
        check_val_every_n_epoch=25,
        limit_train_batches=1.0,
        limit_val_batches=1.0,
        num_sanity_val_steps=2,
        callbacks=RenderCallback(),
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()


```

As you can see from the logs I included, the validation epoch end hooks are called every 25 epochs as specified in the trainer:
```
....
train epoch end on epoch 23
train epoch end on epoch 23
on train epoch end in callback
val epoch end on epoch 24  <--- here
val epoch end on epoch 24  <--- here
on val epoch end in callback  <--- here
train epoch end on epoch 24
train epoch end on epoch 24
on train epoch end in callback
train epoch end on epoch 25
train epoch end on epoch 25
on train epoch end in callback
....
```

Please note that in your render method, you have early returns based on some conditions. Can you please check again that your observations are correct and that you were not just tricked by some missing logs?

**Comment by athn-nik** [comment]
Thanks a lot for your help seems likes there is a mismatch between my epoch indices checks.
